src/compiler/checker.ts

```diff
@@ -28527,37 +28527,33 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             node.kind === SyntaxKind.PropertyDeclaration)!;
     }
 
-    // Check if a parameter or catch variable is assigned anywhere
-    function isSymbolAssigned(symbol: Symbol) {
-        if (!symbol.valueDeclaration) {
-            return false;
-        }
-        const parent = getRootDeclaration(symbol.valueDeclaration).parent;
+    function isSomeSymbolAssigned(rootDeclaration: Node) {
+        const parent = rootDeclaration.parent;
         const links = getNodeLinks(parent);
         if (!(links.flags & NodeCheckFlags.AssignmentsMarked)) {
             links.flags |= NodeCheckFlags.AssignmentsMarked;
             if (!hasParentWithAssignmentsMarked(parent)) {
                 markNodeAssignments(parent);
             }
         }
-        return symbol.isAssigned || false;
+        return getNodeLinks(rootDeclaration).someSymbolAssigned;
     }
 
     function hasParentWithAssignmentsMarked(node: Node) {
         return !!findAncestor(node.parent, node => (isFunctionLike(node) || isCatchClause(node)) && !!(getNodeLinks(node).flags & NodeCheckFlags.AssignmentsMarked));
     }
 
-    function markNodeAssignments(node: Node) {
+    function markNodeAssignments(node: Node): true | undefined {

```

Comment:
Can we not `return` in this function, the way it was before? I think it's not really necessary and potentially confusing, since we don't use the results returned by this function.

Reply:
> since we don't use the results returned by this function.

It is used by `forEachChild`. This `return` acts as an early return for its iteration.

That being said, I probably need to mark both individual symbols as assigned or not together with the aggregate information on the root declaration to fix the regression found [here](https://github.com/microsoft/TypeScript/pull/56313#issuecomment-1841559785).

A minimal repro case for this regression:
```ts
declare function useRouter(): {
    push: (href: string) => void;
}

type SidebarFooterButtonProps = {
  href?: string;
  onClick?: () => void;
};

export const SidebarFooterButton = ({
  href,
  onClick,
}: SidebarFooterButtonProps): JSX.Element => {
  const router = useRouter();

  if (href !== undefined) {
    onClick = () => {
      void router.push(href);
    };
  }

  return (
    <button type="button" onClick={onClick} />
  );
};
```

We still want to narrow "const" parameters and only turn off the dependent parameter narrowing if some of the relevant symbols are reassigned - so we need to keep track of both (or well, derive the aggregate information on demand from the root declaration's symbols)